### ISSUE TITLE
Strict mode does not work for RootModel properties

### ISSUE THREAD

**Issue by Mardoxx** [pydantic/pydantic#10205]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

When a property on a strict BaseModel has a type that inherits from RootModel, the model should enforce strict type validation and not accept any types other than the exact type specified.

### Example Code

```Python
from pydantic import BaseModel, RootModel, ConfigDict

class Foo(RootModel[int]): ...

class TestPermissive(BaseModel):
    model_config = ConfigDict(strict=False)

    a: Foo

class TestStrict(BaseModel):
    model_config = ConfigDict(strict=True)

    a: Foo

TestPermissive(a = 1) # No error (correct)
TestStrict(a = Foo(1)) # No Error (correct)
TestStrict(a = 1) # No Error (incorrect)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.8.2
        pydantic-core version: 2.20.1
          pydantic-core build: profile=release pgo=true
                 install path: /Users/owner/code/codewars/.venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.7 (main, Feb 11 2024, 14:54:11) [Clang 15.0.0 (clang-1500.0.40.1)]
                     platform: macOS-13.5-arm64-arm-64bit
             related packages: typing_extensions-4.12.2
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
@Mardoxx, I disagree, and this would contradict the logic we use with regular models as well:

```py
from pydantic import BaseModel, ConfigDict

class Foo(BaseModel):
    one: int
    two: int

class TestStrict(BaseModel):
    model_config = ConfigDict(strict=True)

    a: Foo

print(TestStrict(a=Foo(one=1, two=2)))
#> a=Foo(one=1, two=2)

print(TestStrict(a={'one': 1, 'two': 2}))
#> a=Foo(one=1, two=2)
```

Marking as not planned, accordingly.

**Comment by Mardoxx** [comment]
Not following how that is comparable. If I had

```
class Foo(BaseModel):
    model_config = ConfigDict(strict=True)
    
    one: str
    two: int

If you stick a `str` into `two` in a strict model it will fail, so it follows that if the property type is a `RootModel` it should only accept that type

**Comment by sydney-runkle** [comment]
I guess my analogy is that a root model is just a wrapper around a type. Thus, we either accept the type or an instance of said root model. Similarly, if a field has a `BaseModel` type, we accept either the type represented via a dict (kind of like the raw type) or an instance of said type.

**Comment by Mardoxx** [comment]
> we either accept the type or an instance of said root model

Providing the (inner) type of the RootModel to a field which is explicitly assigned the RootModel subclass, I would expect to receive only the subclass - not the inner type (they are, after-all, different types)

**Comment by Mardoxx** [comment]
I noticed this also

```
class Foo(RootModel[Annotated[FooType, Strict]]): ...
Foo("1") # passes (incorrect, should fail since not a FooType)
Foo(FooType("1")) # passes (correct)


class Model(RootModel[str]):... # our "Wrapped" str (acts as a str)
SubModel = NewType("SubModel", str) # Equivalent to FooType
class Bar(RootModel[Annotated[SubModel, Strict]]): ...
Bar(SubModel("a")) # Correctly passes
Bar(Model("a")) # Correctly fails, since not a SubModel
```

**Comment by Mardoxx** [comment]
Would there be a way to make it /stricter/ and just not accept anything other than the correct types. This seems more in the spirit of ensuring that it only conforms to the correct schema - not schema-like

**Comment by sydney-runkle** [comment]
> Providing the (inner) type of the RootModel to a field which is explicitly assigned the RootModel subclass, I would expect to receive only the subclass - not the inner type (they are, after-all, different types)

Could you give an example of this?

**Comment by sydney-runkle** [comment]
@Mardoxx,

Could you give a bit more context regarding your use case? Sounds like perhaps you should be using custom types if you want this much control over strict vs lax behavior - the behavior you desire isn't in line with the current `RootModel` behavior, which isn't likely to change (at least in v2) due to an abundance of users currently dependent on these patterns.

**Comment by Mardoxx** [comment]
```py
from pydantic import BaseModel, RootModel

class Foo(RootModel[int]):
    model_config = ConfigDict(strict=True)

class Bar(BaseModel):
    model_config = ConfigDict(strict=True)

    foo: Foo

Bar(foo=1)
```

Interestingly, pyright correctly ascertains that this is wrong:

```py
Argument of type "Literal[1]" cannot be assigned to parameter "foo" of type "Foo" in function "__init__"
  "Literal[1]" is incompatible with "Foo"
```
`Bar(foo=Foo(1))` is correct.


Similarly is the case with NewType
```py
from pydantic import BaseModel
from typing import NewType

Foo = NewType("Foo", int)
class Bar(BaseModel):
    model_config = ConfigDict(strict=True)

    foo: Foo

Bar(foo=1) # This should fail, pyright correctly picks this up as invalid
Bar(foo=Foo(1))
```

**Comment by Mardoxx** [comment]
The example I gave gives context, c.f. https://docs.python.org/3/library/typing.html#newtype

**Comment by sydney-runkle** [comment]
I guess specifically I mean, why not just use the type in these cases, like `int` instead of `Foo`?

**Comment by Mardoxx** [comment]
Let's say you're working with different kinds of id-able entities in your code.. like a UserId, an OrderId, and a ProductId. Without making them specific types, all these would just be plain numbers (integers), but really they are distinct pieces of information. UserId(1) is dissimilar to ProductId(1). This could lead to mistakes where you accidentally use a ProductId where a UserId is expected, or have a less experience engineer use an int -from any source-.

When you create a UserId as a specific type based on an integer, you're telling folk, "This isn't just any number—this number specifically represents a UserId." It helps keep things organised and makes it less prone to errors, because now the type checkers can make sure you're using the right type of id in the right place.

Using specific types like UserId in data contracts is crucial because it ensures that each piece of data is clearly defined and used correctly, reducing the risk of errors and miscommunication between different parts of a system or between systems.

**Comment by Mardoxx** [comment]
Perhaps I could have explained it better, apologies for the lack of clarity

**Comment by Mardoxx** [comment]
@sydney-runkle any update, can we reopen it for visibility also?
